File: LeetCode Daily/02008. Maximum Earnings From Taxi.py
from typing import List
import bisect

class Solution:
    def maxTaxiEarnings(self, n: int, rides: List[List[int]]) -> int:
        if not rides:
            return 0

        # Sort rides by end time
        rides.sort(key=lambda x: x[1])

        # Extract end times for binary search
        end_times = [ride[1] for ride in rides]

        # dp[i] represents the maximum earnings considering rides[0:i+1]
        dp = [0] * len(rides)

        for i in range(len(rides)):
            start, end, tip = rides[i]
            earnings = end - start + tip

            # Option 1: Don't take this ride
            dp[i] = dp[i-1] if i > 0 else 0

            # Option 2: Take this ride
            # Use bisect to find the rightmost ride that ends <= start
            # bisect_right returns the insertion point, so we need index - 1
            idx = bisect.bisect_right(end_times, start, 0, i) - 1

            prev_earnings = dp[idx] if idx >= 0 else 0
            dp[i] = max(dp[i], prev_earnings + earnings)

        return dp[-1]

File: LeetCode Daily/02008. Maximum Earnings From Taxi.go
import "sort"

func maxTaxiEarnings(n int, rides [][]int) int64 {
	if len(rides) == 0 {
		return 0
	}

	// sort rides by end time
	sort.Slice(rides, func(i, j int) bool { return rides[i][1] < rides[j][1] })

	m := len(rides)
	endTimes := make([]int, m)
	for i := 0; i < m; i++ {
		endTimes[i] = rides[i][1]
	}

	dp := make([]int64, m)
	for i := 0; i < m; i++ {
		start := rides[i][0]
		end := rides[i][1]
		tip := rides[i][2]
		earnings := int64(end - start + tip)

		// option 1: skip this ride
		if i > 0 {
			dp[i] = dp[i-1]
		} else {
			dp[i] = 0
		}

		// find rightmost ride that ends <= start among [0, i)
		idx := sort.Search(i, func(j int) bool { return endTimes[j] > start }) - 1

		var prev int64
		if idx >= 0 {
			prev = dp[idx]
		}
		if prev+earnings > dp[i] {
			dp[i] = prev + earnings
		}
	}

	return dp[m-1]
}